### PR TITLE
Set sessionInfo after completing silent auth

### PR DESCRIPTION
**Pull request by Vinnl**
After completing silent authentication in an iframe, the `info` property on the Session wasn't updated. It is now, and I've extracted the code for doing so, which is shared with `handleIncomingRedirect`, into the `setSessionInfo` method, primarily so that there's a single place to change to make sure the session is silently refreshed rather than terminated when it expires.

- [x] I've added a unit test to test for potential regressions of this bug.
- [x] The changelog has been updated, if applicable. N/A
- [x] Commits in this PR are minimal and [have descriptive commit messages](https://chris.beams.io/posts/git-commit/).

### CODE REVIEW

File: packages/browser/src/Session.ts [hunk 1]
```diff
@@ -187,7 +193,16 @@ export class Session extends EventEmitter {
       };
     }
     // Listen for messages from children iframes.
-    setupIframeListener(this.clientAuthentication.handleIncomingRedirect);
+    setupIframeListener(async (redirectUrl: string) => {
+      const sessionInfo = await this.clientAuthentication.handleIncomingRedirect(
+        redirectUrl
+      );
+
+      if (!isLoggedIn(sessionInfo)) {
```
Comment (pmcb55) [inline packages/browser/src/Session.ts:201]: If it's possible for this handler to be invoked when already logged in, then should we still call `setSessionInfo()` regardless, i.e., maybe we were called because we want to reset our session's access token timer to a new expiration date?
Or to state this another way, is there a very specific reason that we don't want to call `setSessionInfo()`? (If so, then a comment is warranted perhaps?)

Comment (Vinnl) [reply packages/browser/src/Session.ts:201]: That's a good question. @NSeydoux, do you know the answer? 'cause I don't...

File: packages/browser/src/Session.ts [hunk 2]
```diff
@@ -424,4 +431,17 @@ export class Session extends EventEmitter {
...
     this.on("sessionRestore", callback);
   }
+
+  private setSessionInfo(
+    sessionInfo: ISessionInfo & { isLoggedIn: true }
+  ): void {
+    this.info.isLoggedIn = sessionInfo.isLoggedIn;
+    this.info.webId = sessionInfo.webId;
+    this.info.sessionId = sessionInfo.sessionId;
+    if (typeof sessionInfo.expirationDate === "number") {
```
Comment (NSeydoux) [inline packages/browser/src/Session.ts:441]:
```suggestion
    this.info.sessionId = sessionInfo.sessionId;
    this.info.expirationDate = sessionInfo.expirationDate;
    if (typeof sessionInfo.expirationDate === "number") {
```
This will allow, in an upcoming ticket, to change the behaviour of the callback set by `setTimeout` to checking if the current expiration date (`this.info.expirationDate`) is close to `Date.now()` (with a certain tolerance). This way, if silent authentication happens in an `iframe`, the callback will not log the session out.

File: packages/browser/__tests__/Session.spec.ts [hunk 3]
```diff
@@ -1020,5 +1020,74 @@ describe("Session", () => {
...
+        redirectUrl: "https://some.redirect/url",
+        tokenType: "DPoP",
+      });
+      clientAuthentication.validateCurrentSession = jest
+        .fn()
+        .mockReturnValue(validateCurrentSessionPromise);
+      const incomingRedirectPromise = Promise.resolve({
+        isLoggedIn: true,
+        webId: "https://some.pod/profile#me",
+        sessionId: "someSessionId",
```
Comment (NSeydoux) [inline packages/browser/__tests__/Session.spec.ts:1054]:
```suggestion
        sessionId: "someSessionId",
        expirationDate: 961106400
```
This sets a new expiration date for the access token held by the session. It happens to be the release date of Pokemon yellow in Europe, arguably an excellent day, but that's purely coincidental.

File: packages/browser/__tests__/Session.spec.ts [hunk 4]
```diff
@@ -1020,5 +1020,74 @@ describe("Session", () => {
...
+        },
+      } as MessageEvent;
+      // This handler will call `clientAuthentication.handleIncomingRedirect`,
+      // which will return our sessionInfo values:
+      messageEventHandler(mockedEvent);
+
+      await incomingRedirectPromise;
+      await validateCurrentSessionPromise;
+      expect(mySession.info.webId).toBe("https://some.pod/profile#me");
+      expect(mySession.info.sessionId).toBe("someSessionId");
```
Comment (NSeydoux) [inline packages/browser/__tests__/Session.spec.ts:1090]:
```suggestion
      expect(mySession.info.sessionId).toBe("someSessionId");
      expect(mySession.info.expirationDate).toBe(961106400);
```
I'm repeating myself but... what a date !

File: packages/browser/__tests__/Session.spec.ts [hunk 5]
```diff
@@ -1020,5 +1020,74 @@ describe("Session", () => {
...
+        data: {
+          redirectUrl: "http://arbitrary.com",
+        },
+      } as MessageEvent;
+      // This handler will call `clientAuthentication.handleIncomingRedirect`,
+      // which will return our sessionInfo values:
+      messageEventHandler(mockedEvent);
+
+      await incomingRedirectPromise;
+      await validateCurrentSessionPromise;
```
Comment (NSeydoux) [inline packages/browser/__tests__/Session.spec.ts:1088]:
```suggestion
```

File: packages/browser/src/Session.ts [hunk 6]
```diff
@@ -187,7 +193,16 @@ export class Session extends EventEmitter {
...
-    setupIframeListener(this.clientAuthentication.handleIncomingRedirect);
+    setupIframeListener(async (redirectUrl: string) => {
+      const sessionInfo = await this.clientAuthentication.handleIncomingRedirect(
+        redirectUrl
+      );
+
+      if (!isLoggedIn(sessionInfo)) {
+        return;
+      }
+      this.setSessionInfo(sessionInfo);
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply packages/browser/src/Session.ts]: Yes, because I'd actually forgotten as well: 4dd2b96c

File: packages/browser/__tests__/Session.spec.ts [hunk 7]
```diff
@@ -1020,5 +1020,74 @@ describe("Session", () => {
         inIframe: true,
       });
     });
+
+    it("sets the updated session info after silently refreshing", async () => {
+      const sessionId = "mySession";
+      mockLocalStorage({
+        [KEY_CURRENT_SESSION]: sessionId,
+      });
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply packages/browser/__tests__/Session.spec.ts]: Good point, I was kind of intimidated by all the mocking ceremony in the other test, but I should've checked that actually most of it isn't needed for this test: 0b4e6ddc

File: packages/browser/__tests__/Session.spec.ts [hunk 8]
```diff
@@ -1020,5 +1020,74 @@ describe("Session", () => {
...
+      const validateCurrentSessionPromise = Promise.resolve({
+        issuer: "https://some.issuer",
+        clientAppId: "some client ID",
+        clientAppSecret: "some client secret",
+        redirectUrl: "https://some.redirect/url",
+        tokenType: "DPoP",
+      });
+      clientAuthentication.validateCurrentSession = jest
+        .fn()
+        .mockReturnValue(validateCurrentSessionPromise);
```
[the remark this answers is not in the document]
Comment (Vinnl) [reply packages/browser/__tests__/Session.spec.ts]: Good point, I was kind of intimidated by all the mocking ceremony in the other test, but I should've checked that actually most of it isn't needed for this test: 0b4e6ddc

File: packages/browser/src/Session.ts (hunk 2 above)
Comment (Vinnl) [reply packages/browser/src/Session.ts]: Whoops yeah, that's kinda important: 4134dc9f

File: packages/browser/__tests__/Session.spec.ts (hunk 3 above)
Comment (Vinnl) [reply packages/browser/__tests__/Session.spec.ts:1029]: 4134dc9f

File: packages/browser/__tests__/Session.spec.ts (hunk 4 above)
Comment (Vinnl) [reply packages/browser/__tests__/Session.spec.ts:1064]: 4134dc9f

File: packages/browser/__tests__/Session.spec.ts (hunk 5 above)
Comment (Vinnl) [reply packages/browser/__tests__/Session.spec.ts]: 0b4e6ddc

File: packages/browser/src/Session.ts (hunk 1 above)
Comment (NSeydoux) [reply packages/browser/src/Session.ts:201]: Well technically, the redirection failing does not mean that the session isn't logged out. Imagine that a user initially logs in an gets an access token valid for 30 minutes. If we try to perform a silent authentication in an iframe 10 minutes before the expiration, and it fails, then the session isn't logged out for another 10 minutes, at which point the callback we set initially when receiving the token fires and the session gets properly logged out, emitting a logout signal, clearing the storage and all. This is exaggerated, but the idea is that if the silent authentication is successful, we want to update the session info in particular with the new expiration date, but if it failed we don't do anything and have dedicated code properly logging out the session.

So long story short I think the current implementation is the "right" thing to do, but we may add a comment specifying that the proper logout happens in a callback set when receiving the access token.

Comment (Vinnl) [reply packages/browser/src/Session.ts:204]: Makes sense, thanks! Added comment: 0d15be9a

File: packages/browser/src/Session.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (NSeydoux) [reply packages/browser/src/Session.ts]: "Your comments are for your future self" is one of the best pieces of advice I've had from a CS teacher :D

File: packages/browser/__tests__/Session.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (NSeydoux) [reply packages/browser/__tests__/Session.spec.ts]: That's fair, setting up the mocks where there are so much side effects is kind of involved
